kinesis/src/main/scala/akka/stream/alpakka/kinesisfirehose/scaladsl/KinesisFirehoseFlow.scala

```diff
@@ -5,37 +5,44 @@
 package akka.stream.alpakka.kinesisfirehose.scaladsl
 
 import akka.NotUsed
+import akka.dispatch.ExecutionContexts.sameThreadExecutionContext
 import akka.stream.ThrottleMode
 import akka.stream.alpakka.kinesisfirehose.KinesisFirehoseFlowSettings
-import akka.stream.alpakka.kinesisfirehose.impl.KinesisFirehoseFlowStage
+import akka.stream.alpakka.kinesisfirehose.KinesisFirehoseErrors.FailurePublishingRecords
 import akka.stream.scaladsl.Flow
-import com.amazonaws.services.kinesisfirehose.AmazonKinesisFirehoseAsync
-import com.amazonaws.services.kinesisfirehose.model.{PutRecordBatchResponseEntry, Record}
+import software.amazon.awssdk.services.firehose.FirehoseAsyncClient
+import software.amazon.awssdk.services.firehose.model.{PutRecordBatchRequest, PutRecordBatchResponseEntry, Record}
 
 import scala.collection.JavaConverters._
 import scala.collection.immutable.Queue
 import scala.concurrent.duration._
 
+import scala.compat.java8.FutureConverters._
+
 object KinesisFirehoseFlow {
   def apply(streamName: String, settings: KinesisFirehoseFlowSettings = KinesisFirehoseFlowSettings.Defaults)(
-      implicit kinesisClient: AmazonKinesisFirehoseAsync
+      implicit kinesisClient: FirehoseAsyncClient
   ): Flow[Record, PutRecordBatchResponseEntry, NotUsed] =
     Flow[Record]
       .throttle(settings.maxRecordsPerSecond, 1.second, settings.maxRecordsPerSecond, ThrottleMode.Shaping)
       .throttle(settings.maxBytesPerSecond, 1.second, settings.maxBytesPerSecond, getByteSize, ThrottleMode.Shaping)
       .batch(settings.maxBatchSize, Queue(_))(_ :+ _)
-      .via(
-        new KinesisFirehoseFlowStage(
-          streamName,
-          settings.maxRetries,
-          settings.backoffStrategy,
-          settings.retryInitialTimeout
-        )
+      .mapAsync(settings.parallelism)(
+        records =>
+          kinesisClient
+            .putRecordBatch(
+              PutRecordBatchRequest
+                .builder()
+                .deliveryStreamName(streamName)
+                .records(records.asJavaCollection)
+                .build()
+            )
+            .toScala
+            .transform(identity, FailurePublishingRecords(_))(sameThreadExecutionContext)
       )
-      .mapAsync(settings.parallelism)(identity)
-      .mapConcat(_.getRequestResponses.asScala.toIndexedSeq)
-      .filter(_.getErrorCode == null)
+      .mapConcat(_.requestResponses.asScala.toIndexedSeq)
+      .filter(_.errorCode == null)
```

Comment:
This is interesting, are elements with errors silently discarded?

Reply:
Ha, I failed to remove that line, I think now that it's all handled by the AWS client, it is expected users will either intentionally set up a retry strategy or simply be oblivious to it. In both cases, errors should not be discarded. Thanks for the catch 👍